common/source/nrt_fpga_device.cpp

```diff
@@ -0,0 +1,369 @@
+				&imu_value.magx, sizeof(short));
+		IOFPGARead(nrt_fpga_fd_, IOFPGA_IMU_VALUE_ADDR(imu_data_addr, magy), \
+				&imu_value.magy, sizeof(short));
+		IOFPGARead(nrt_fpga_fd_, IOFPGA_IMU_VALUE_ADDR(imu_data_addr, magz), \
+				&imu_value.magz, sizeof(short));
+
+		result.resize((sizeof(struct st_iofpga_imu_value)));
+		result[0] = imu_value.accx;
+		result[1] = imu_value.accx >> 8 & 0xff;
+		result[2] = imu_value.accy;
+		result[3] = imu_value.accy >> 8 & 0xff;
+		result[4] = imu_value.accz;
+		result[5] = imu_value.accz >> 8 & 0xff;
+
+		result[6] = imu_value.gyrox;
+		result[7] = imu_value.gyrox >> 8 & 0xff;
+		result[8] = imu_value.gyroy;
+		result[9] = imu_value.gyroy >> 8 & 0xff;
+		result[10] = imu_value.gyroz;
+		result[11] = imu_value.gyroz >> 8 & 0xff;
+
+		result[12] = imu_value.magx;
+		result[13] = imu_value.magx >> 8 & 0xff;
+		result[14] = imu_value.magy;
+		result[15] = imu_value.magy >> 8 & 0xff;
+		result[16] = imu_value.magz;
+		result[17] = imu_value.magz >> 8 & 0xff;
+	}
+	else
+		return kInvalidArgument;
+
+	return kNoError;
+}
+
+error_condition NRTFPGADevice::Set(const string& key, \
+									const vector<uint8_t>& value) {
+	static const error_condition kInvalidArgument(errc::invalid_argument);
+	if (key.empty())
+		return kInvalidArgument;
+	if (key == kDOData) {
+		short do_p_data = 0;
+		short do_h_data = 0;
+		if (value.size() < sizeof(do_p_data) + sizeof(do_h_data))
+			return kInvalidArgument;
+
+		do_p_data = value[0] & (value[1] << 8 & 0xff00);
+		do_h_data = value[2];
+		IOFPGAWrite(nrt_fpga_fd_, \
+				IOFPGA_DO_DATA_ADDR(info_.driver._do->data_offset, pdata),\
+				do_p_data, \
+				sizeof(do_p_data));
+		IOFPGAWrite(nrt_fpga_fd_, \
+				IOFPGA_DO_DATA_ADDR(info_.driver._do->data_offset, hdata),\
+				do_h_data, \
+				sizeof(do_h_data));
+	}
+	else
+		return kInvalidArgument;
+
+	return kNoError;
+}
+
+error_condition NRTFPGADevice::IOFPGAWrite(int fd, int addr, long long value, \
+		int size) {
+	int result;
+	if (size < 1 || size > 8)
+		return make_error_condition(errc::invalid_argument);
+
+	ioctl(fd, IOFPGA_WR_ADDR, addr);
+	result = write(fd, &value, size);
+	if (result < 0)
+		return make_error_condition(errc::io_error);
+
+	return kNoError;
+}
+
+error_condition NRTFPGADevice::IOFPGARead(int fd, int addr, void * buffer, \
+		int size)
+{
+	int result;
+	ioctl(fd, IOFPGA_WR_ADDR, addr);
+	result = read(fd, buffer, size);
+	if (result < 0)
+		return make_error_condition(errc::io_error);
+
+	return kNoError;
+}
+
+error_condition NRTFPGADevice::IOFPGAGetInfo(int fd) {
+	void **driver_info, ** hardware_info;
+
+	ioctl(fd, IOFPGA_SEL_INFO, NULL);
+	IOFPGARead(fd, 0, &info_page_, sizeof(info_page_));
+
+	memset(&info_, 0, sizeof(info_));
+
+	info_.hardware.chip = (struct st_iofpga_info *)info_page_.data;
+
+	for (int i = 0; i < info_.hardware.chip->fcount; i++)
+	{
+		switch (info_.hardware.chip->funcs[i].id)
+		{
+			case IOFPGA_HW_FUNCID_CLK:
+				driver_info = (void **)&info_.driver.clk;
```

Comment:
Maybe using some kind of cast on the type conversion, like `static_cast` (may not be working) or `reinterpret_cast`.